File: java-dci/src/example/domain/TransferMoneySource.java
package example.domain;

import example.dci.Actor;
import example.dci.Role;

/**
 * Methodful role to implement how Money is transferred from a Source
 * 
 * @author Raz
 *
 */
public class TransferMoneySource implements MoneySource {
	Account account;
	
	@Override
	public Role apply(Actor actor) {
		this.account = (Account) actor;
		return this;
	}

	@Override
	public void withdraw(double amount) {
		if (account.getBalance() > amount) {
			account.decreaseBalance(amount);
		} else {
			throw new IllegalArgumentException("Insufficient Balance in Source");
		}
	}
}

File: scala-dci/src/example/domain/roles/TransferMoneySource.scala
package example.domain.roles

import example.domain.Account

/**
 * This is the methodful role for the source account for the money transfer
 */
trait TransferMoneySource extends MoneySource { 
  this: Account =>
  def transfer(amount: Long, recipient: MoneyRecipient) {
    if (availableBalance < amount) { 
      throw new IllegalArgumentException
    } else {
      decreaseBalance(amount)
      recipient.increaseBalance(amount)
    }
  }
}
